File: Flutter/android/src/main/java/com/tencent/cloud/tuikit/flutter/tuicallkit/view/incomingfloatwindow/IncomingFloatView.java
package com.tencent.cloud.tuikit.flutter.tuicallkit.view.incomingfloatwindow;

import android.content.Context;
import android.graphics.PixelFormat;
import android.net.Uri;
import android.os.Build;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.tencent.cloud.tuikit.engine.call.TUICallDefine;
import com.tencent.cloud.tuikit.engine.call.TUICallEngine;
import com.tencent.cloud.tuikit.flutter.tuicallkit.R;
import com.tencent.cloud.tuikit.flutter.tuicallkit.TUICallKitPlugin;
import com.tencent.cloud.tuikit.flutter.tuicallkit.state.TUICallState;
import com.tencent.cloud.tuikit.flutter.tuicallkit.state.User;
import com.tencent.cloud.tuikit.flutter.tuicallkit.utils.Constants;
import com.tencent.cloud.tuikit.flutter.tuicallkit.utils.Permission;
import com.tencent.cloud.tuikit.flutter.tuicallkit.view.WindowManager;
import com.tencent.cloud.tuikit.flutter.tuicallkit.utils.Logger;
import com.tencent.qcloud.tuicore.TUICore;
import com.tencent.qcloud.tuicore.permission.PermissionRequester;
import com.tencent.qcloud.tuicore.util.ScreenUtil;

public class IncomingFloatView {
    private static final String                                  TAG = "IncomingFloatView";
    private              android.view.WindowManager              mWindowManager;
    private              android.view.WindowManager.LayoutParams mWindowLayoutParams;

    private View      layoutView;
    private ImageView imageFloatAvatar;
    private TextView  textFloatTitle;
    private TextView  textFloatDescription;
    private ImageView imageReject;
    private ImageView imageAccept;

    private User                    user;
    private TUICallDefine.MediaType mediaType;

    private Context context;

    private int padding = 20;

    public IncomingFloatView(Context context) {
        this.context = context;
    }

    public void showIncomingView(User caller, TUICallDefine.MediaType mediaType) {
        Logger.info(TUICallKitPlugin.TAG, "IncomingFloatView  showIncomingView | UserId:" + caller.id);

        this.user = caller;
        this.mediaType = mediaType;
        initWindow();
    }

    private void initWindow() {
        Logger.info(TUICallKitPlugin.TAG, "IncomingFloatView  initWindow");

        layoutView = LayoutInflater.from(context).inflate(R.layout.tuicallkit_incoming_float_view, null);
        imageFloatAvatar = layoutView.findViewById(R.id.img_float_avatar);
        textFloatTitle = layoutView.findViewById(R.id.tv_float_title);
        textFloatDescription = layoutView.findViewById(R.id.tv_float_desc);
        imageReject = layoutView.findViewById(R.id.btn_float_decline);
        imageAccept = layoutView.findViewById(R.id.btn_float_accept);

        Uri avatarUri = Uri.parse(user.avatar);
        if (avatarUri == null) {
            if (imageFloatAvatar != null && R.drawable.tuicallkit_ic_avatar != 0) {
                imageFloatAvatar.setImageResource(R.drawable.tuicallkit_ic_avatar);
            }
        } else {
            Glide.with(context).load(avatarUri).error(R.drawable.tuicallkit_ic_avatar).into(imageFloatAvatar);
        }

        textFloatTitle.setText(user.nickname);
        if (mediaType == TUICallDefine.MediaType.Video) {
            String str =
                    (String) TUICallState.getInstance().mResourceMap.get(TUICallState.getInstance().mScene == TUICallDefine.Scene.SINGLE_CALL ? "k_0000002_1" : "k_0000003");
            textFloatDescription.setText(str);

            imageAccept.setImageResource(R.drawable.tuicallkit_ic_dialing_video);
        } else {
            String str =
                    (String) TUICallState.getInstance().mResourceMap.get(TUICallState.getInstance().mScene == TUICallDefine.Scene.SINGLE_CALL ? "k_0000002" : "k_0000003");
            textFloatDescription.setText(str);

            imageAccept.setImageResource(R.drawable.tuicallkit_ic_dialing);
        }
        imageReject.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                cancelIncomingView();
                TUICallEngine.createInstance(context).reject(null);
            }
        });

        layoutView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                cancelIncomingView();
                if (Permission.hasPermission(PermissionRequester.BG_START_PERMISSION)) {
                    WindowManager.launchMainActivity(context);
                }
                TUICore.notifyEvent(Constants.KEY_CALLKIT_PLUGIN, Constants.SUB_KEY_HANDLE_CALL_RECEIVED, null);
            }
        });

        imageAccept.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                cancelIncomingView();
                TUICallEngine.createInstance(context).accept(null);
                if (Permission.hasPermission(PermissionRequester.BG_START_PERMISSION)) {
                    WindowManager.launchMainActivity(context);
                }
                TUICore.notifyEvent(Constants.KEY_CALLKIT_PLUGIN, Constants.SUB_KEY_HANDLE_CALL_RECEIVED, null);
            }
        });

        mWindowManager = (android.view.WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        mWindowManager.addView(layoutView, getViewParams());
    }

    private android.view.WindowManager.LayoutParams getViewParams() {
        mWindowLayoutParams = new android.view.WindowManager.LayoutParams();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            mWindowLayoutParams.type = android.view.WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY;
        } else {
            mWindowLayoutParams.type = android.view.WindowManager.LayoutParams.TYPE_PHONE;
        }
        mWindowLayoutParams.flags = android.view.WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL
                | android.view.WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
                | android.view.WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS;

        mWindowLayoutParams.gravity = Gravity.START | Gravity.TOP;
        mWindowLayoutParams.x = padding;
        mWindowLayoutParams.y = 0;

        mWindowLayoutParams.width = ScreenUtil.getRealScreenWidth(context) - padding * 2;
        mWindowLayoutParams.height = android.view.WindowManager.LayoutParams.WRAP_CONTENT;
        mWindowLayoutParams.format = PixelFormat.TRANSPARENT;

        return mWindowLayoutParams;
    }

    public void cancelIncomingView() {
        Logger.info(TUICallKitPlugin.TAG, "IncomingFloatView  cancelIncomingView");

        if (layoutView != null && layoutView.isAttachedToWindow() && mWindowManager != null) {
            mWindowManager.removeView(layoutView);
        }
        layoutView = null;
    }
}

File: Android/tuicallkit-kt/src/main/java/com/tencent/qcloud/tuikit/tuicallkit/view/component/incomingview/IncomingFloatView.kt
package com.tencent.qcloud.tuikit.tuicallkit.view.component.incomingview

import android.content.Context
import android.content.Context.WINDOW_SERVICE
import android.content.res.Configuration
import android.graphics.PixelFormat
import android.os.Build
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.WindowManager
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import com.tencent.cloud.tuikit.engine.call.TUICallDefine
import com.tencent.qcloud.tuicore.TUICore
import com.tencent.qcloud.tuicore.interfaces.ITUINotification
import com.tencent.qcloud.tuicore.permission.PermissionCallback
import com.tencent.qcloud.tuicore.util.ScreenUtil
import com.tencent.qcloud.tuikit.tuicallkit.R
import com.tencent.qcloud.tuikit.tuicallkit.data.Constants
import com.tencent.qcloud.tuikit.tuicallkit.data.User
import com.tencent.qcloud.tuikit.tuicallkit.manager.EngineManager
import com.tencent.qcloud.tuikit.tuicallkit.state.TUICallState
import com.tencent.qcloud.tuikit.tuicallkit.utils.ImageLoader
import com.tencent.qcloud.tuikit.tuicallkit.utils.Logger
import com.tencent.qcloud.tuikit.tuicallkit.utils.PermissionRequest
import com.tencent.qcloud.tuikit.tuicallkit.view.component.videolayout.VideoViewFactory
import com.trtc.tuikit.common.livedata.Observer

class IncomingFloatView(context: Context) : RelativeLayout(context) {
    companion object {
        private const val TAG = "IncomingViewFloat"
    }

    private var appContext: Context = context.applicationContext
    private var caller: User? = null

    private var windowManager: WindowManager? = null
    private var windowLayoutParams: WindowManager.LayoutParams? = null

    private lateinit var layoutView: View
    private var imageFloatAvatar: ImageView? = null
    private var textFloatTitle: TextView? = null
    private var textFloatDescription: TextView? = null
    private var imageReject: ImageView? = null
    private var imageAccept: ImageView? = null

    private val padding = 40

    private var callStatusObserver = Observer<TUICallDefine.Status> {
        if (it == TUICallDefine.Status.None || it == TUICallDefine.Status.Accept) {
            cancelIncomingView()
        }
    }

    private val notification = ITUINotification { key, subKey, param ->
        if (key == Constants.EVENT_VIEW_STATE_CHANGED &&
            (subKey == Constants.EVENT_SHOW_FULL_VIEW || subKey == Constants.EVENT_SHOW_FLOAT_VIEW)) {
            cancelIncomingView()
        }
    }

    fun showIncomingView(user: User) {
        Logger.info(TAG, "showIncomingView, user: $user")
        caller = user
        initWindow()
        addObserver()
    }

    fun cancelIncomingView() {
        Logger.info(TAG, "cancelIncomingView")
        if (layoutView.isAttachedToWindow) {
            windowManager?.removeView(layoutView)
        }
        removeObserver()
    }

    private fun addObserver() {
        TUICallState.instance.selfUser.get().callStatus.observe(callStatusObserver)
        TUICore.registerEvent(Constants.EVENT_VIEW_STATE_CHANGED, Constants.EVENT_SHOW_FULL_VIEW, notification)
        TUICore.registerEvent(Constants.EVENT_VIEW_STATE_CHANGED, Constants.EVENT_SHOW_FLOAT_VIEW, notification)
    }

    private fun removeObserver() {
        TUICallState.instance.selfUser.get().callStatus.removeObserver(callStatusObserver)
        TUICore.unRegisterEvent(notification)
    }

    private fun initWindow() {
        layoutView = LayoutInflater.from(context).inflate(R.layout.tuicallkit_incoming_float_view, this)
        imageFloatAvatar = layoutView.findViewById(R.id.img_float_avatar)
        textFloatTitle = layoutView.findViewById(R.id.tv_float_title)
        textFloatDescription = layoutView.findViewById(R.id.tv_float_desc)
        imageReject = layoutView.findViewById(R.id.btn_float_decline)
        imageAccept = layoutView.findViewById(R.id.btn_float_accept)

        ImageLoader.loadImage(appContext, imageFloatAvatar, caller?.avatar?.get(), R.drawable.tuicallkit_ic_avatar)
        textFloatTitle?.text = caller?.nickname?.get()

        textFloatDescription?.text = if (TUICallState.instance.mediaType.get() == TUICallDefine.MediaType.Video) {
            appContext.resources.getString(R.string.tuicallkit_invite_video_call)
        } else {
            appContext.resources.getString(R.string.tuicallkit_invite_audio_call)
        }

        imageReject?.setOnClickListener {
            EngineManager.instance.reject(null)
            cancelIncomingView()
        }

        layoutView.setOnClickListener {
            cancelIncomingView()
            TUICore.notifyEvent(Constants.EVENT_TUICALLKIT_CHANGED, Constants.EVENT_START_ACTIVITY, HashMap())
        }

        if (TUICallState.instance.mediaType.get() == TUICallDefine.MediaType.Video) {
            imageAccept?.setBackgroundResource(R.drawable.tuicallkit_ic_dialing_video)
        } else {
            imageAccept?.setBackgroundResource(R.drawable.tuicallkit_bg_dialing)
        }
        imageAccept?.setOnClickListener {
            if (TUICallState.instance.selfUser.get().callStatus.get() == TUICallDefine.Status.None) {
                Logger.warn(TAG, "current status is None, ignore")
                cancelIncomingView()
                return@setOnClickListener
            }

            PermissionRequest.requestPermissions(appContext, TUICallState.instance.mediaType.get(),
                object : PermissionCallback() {
                    override fun onGranted() {
                        if (TUICallState.instance.selfUser.get().callStatus.get() == TUICallDefine.Status.None) {
                            Logger.warn(TAG, "current status is None, ignore")
                            cancelIncomingView()
                            return
                        }
                        Logger.info(TAG, "accept the call")
                        TUICore.notifyEvent(
                            Constants.EVENT_TUICALLKIT_CHANGED, Constants.EVENT_START_ACTIVITY, HashMap()
                        )
                        EngineManager.instance.accept(null)
                        if (TUICallState.instance.mediaType.get() == TUICallDefine.MediaType.Video) {
                            val videoView = VideoViewFactory.instance.createVideoView(
                                TUICallState.instance.selfUser.get(), appContext
                            )

                            EngineManager.instance.openCamera(
                                TUICallState.instance.isFrontCamera.get(), videoView?.getVideoView(), null
                            )
                        }
                        cancelIncomingView()
                    }

                    override fun onDenied() {
                        super.onDenied()
                        EngineManager.instance.reject(null)
                        cancelIncomingView()
                    }
                })
        }

        windowManager = appContext.getSystemService(WINDOW_SERVICE) as WindowManager
        windowManager?.addView(layoutView, viewParams)
    }

    private val viewParams: WindowManager.LayoutParams
        get() {
            windowLayoutParams = WindowManager.LayoutParams()
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                windowLayoutParams!!.type = WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
            } else {
                windowLayoutParams!!.type = WindowManager.LayoutParams.TYPE_PHONE
            }
            windowLayoutParams!!.flags = (WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL
                    or WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
                    or WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS)
            windowLayoutParams!!.gravity = Gravity.START or Gravity.TOP
            windowLayoutParams!!.x = padding
            windowLayoutParams!!.y = 0
            windowLayoutParams!!.width = ScreenUtil.getScreenWidth(appContext) - padding * 2
            windowLayoutParams!!.height = WindowManager.LayoutParams.WRAP_CONTENT
            windowLayoutParams!!.format = PixelFormat.TRANSPARENT
            return windowLayoutParams as WindowManager.LayoutParams
        }

    override fun onConfigurationChanged(newConfig: Configuration?) {
        super.onConfigurationChanged(newConfig)
        layoutView?.let { windowManager?.updateViewLayout(layoutView, viewParams) }
    }
}